### PR TITLE
Partial Collapse of Multidimensional Auxcoords with Bounds

### PR DESCRIPTION
**Pull request by rcomer**
Proposed fix for #3298.

### CODE REVIEW

File: lib/iris/coords.py [hunk 1]
```diff
@@ -1356,16 +1356,22 @@ def serialize(x):
...
-            # Determine the array library for stacking
-            al = da if self.has_bounds() \
-                and _lazy.is_lazy_data(self.core_bounds()) else np
+            if self.has_bounds():
+                item = self.core_bounds()
+                if dims_to_collapse is not None:
+                    # Express main dims_to_collapse as non-negative integers
+                    # and add the last (bounds specific) dimension.
+                    dims_to_collapse = tuple(
+                        dim % self.ndim for dim in dims_to_collapse) + (-1,)
```
Comment (lbdreyer) [inline lib/iris/coords.py]: Nice!

Comment (lbdreyer) [inline lib/iris/coords.py]: Nice!

File: lib/iris/tests/unit/coords/test_Coord.py [hunk 2]
```diff
@@ -303,6 +303,30 @@ def test_numeric_nd(self):
...
+                                                                [78, 112]]))
+
+        # ... and the other
+        collapsed_coord = coord.collapsed(0)
+        self.assertArrayEqual(
+            collapsed_coord.points, np.array([40, 50, 60, 70]))
+        self.assertArrayEqual(collapsed_coord.bounds, np.array([[-2, 82],
+                                                                [8,  92],
+                                                                [18, 102],
+                                                                [28, 112]]))
```
Comment (lbdreyer) [inline lib/iris/tests/unit/coords/test_Coord.py]: Could you split this into 3 tests? i.e. one test for each collapse operation.

My concern with having them all in the one test is that if the first one fails we won't know the result of the other two collapses until we get the first one working.
I realise `test_numeric_nd()` that already exists in this file does the same thing as you have done here, but I do think it's better practice to split the tests up

Comment (lbdreyer) [reply lib/iris/tests/unit/coords/test_Coord.py]: And perhaps add a test where the dimensions given are negative to test that handling is correct:
i.e. 
```
collapsed_coord = coord.collapsed(-1)
```

Comment (rcomer) [reply lib/iris/tests/unit/coords/test_Coord.py]: Will do, and thanks for explaining why - it makes sense.  Should I do the same with the lazy equivalent below?

Comment (lbdreyer) [reply lib/iris/tests/unit/coords/test_Coord.py]: > Should I do the same with the lazy equivalent below?

Yes please!

Comment (rcomer) [reply lib/iris/tests/unit/coords/test_Coord.py]: Done.  Also moved the dask import to the top of the module.  This is consistent with other test modules that use dask.

Since I now have `self.setupTestArrays((3, 4))` in 8 of the tests, I wondered about adding a setUp method to contain that.  But there are a lot of tests on the class that don't need these arrays so not really sure what the better approach is.

**Comment by rcomer** [comment]
This change also (at least partly) addresses #3363.  I've added an extra test at #3364 that should pass once this branch is merged.

**Comment by lbdreyer** [comment]
All looks good! Thanks @rcomer !

I'm going to merge this in. 
Note the travis tests will fail due to the numpy issues that are fixed in #3366 but I don't want that to hold this PR up
